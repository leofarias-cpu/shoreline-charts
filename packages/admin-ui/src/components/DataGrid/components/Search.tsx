import React from 'react'

import type { SearchProps } from '../../Search'
import { Search as BaseSearch } from '../../Search'

export function Search(props: SearchProps) {
  const { containerCsx, ...searchProps } = props

  return (
    <BaseSearch
      containerCsx={{ width: '25%', minWidth: '12.25rem', ...containerCsx }}
      {...searchProps}
    />
  )
}
