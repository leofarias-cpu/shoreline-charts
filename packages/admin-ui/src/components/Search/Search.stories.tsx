import React, { useState } from 'react'
import type { Story, Meta } from '@storybook/react'

import type { SearchProps } from './index'
import { Search } from './index'

export default {
  title: 'admin-ui/Search',
  component: Search,
} as Meta

export const Playground: Story<SearchProps> = (args) => {
  const [value, setValue] = useState('')

  return (
    <Search
      value={value}
      onChange={(e) => setValue(e.target.value)}
      onClear={() => {
        setValue('')
      }}
      {...args}
    />
  )
}

Playground.args = {
  id: 'search',
  placeholder: 'Search for a product, category or brand',
  onSubmit: () => alert('submited'),
  loading: false,
}
