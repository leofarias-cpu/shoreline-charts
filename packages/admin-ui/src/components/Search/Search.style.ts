import type { StyleProp } from '@vtex/admin-ui-core'

const style: Record<string, StyleProp> = {}

style.input = {
  height: '2.5rem',
  paddingY: '0.4375rem',
  paddingLeft: '2rem',
  paddingRight: '2.5rem',
  margin: 0,
}

style.icon = {
  top: 0,
  margin: '0.625rem 0.25rem 0 0.5rem',
}

style.spinner = {
  ...style.icon,
  position: 'absolute',
}

style.search = {
  ...style.icon,
  color: 'base',
}

style.clearButton = {
  position: 'absolute',
  top: 0,
  right: 0,
  marginRight: '0.375rem',
}

export default style
