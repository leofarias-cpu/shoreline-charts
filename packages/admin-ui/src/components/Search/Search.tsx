import type { ComponentPropsWithRef } from 'react'
import React, { Fragment } from 'react'
import { jsx } from '@vtex/admin-ui-react'
import { IconSearch, IconCancel } from '@vtex/admin-ui-icons'

import type { SearchFormState } from './hooks/useSearchState'
import { AbstractInput } from '../AbstractInput'
import { Button } from '../Button'
import { VisuallyHidden } from '../VisuallyHidden'
import { Label } from '../Label'
import { Spinner } from '../Spinner'

/**
 * Search form
 * @example
 * const state = useSearchState()
 *
 * <Search
 *  state={state}
 *  placeholder="Placeholder"
 *  aria-label="Search"
 * />
 *
 * @link [For accessibility guidelines](https://admin-ui-docs.vercel.app/form/search/#accessibility)
 */
export const Search = jsx('form')(
  {
    position: 'relative',
    marginX: '1px',
  },
  {
    options: ['state'],
    useOptions(options: SearchOptions, props) {
      const { state } = options
      const { id, placeholder, 'aria-label': label = '', ...formProps } = props

      return {
        ...formProps,
        onSubmit: state.onSubmit,
        'aria-label': `${label} Form`,
        role: 'search',
        children: (
          <Fragment>
            <VisuallyHidden>
              <Label htmlFor={id}>{label} Input</Label>
            </VisuallyHidden>
            <AbstractInput
              placeholder={placeholder}
              id={id}
              value={state.value}
              icon={
                state.loading ? (
                  <Spinner
                    size={20}
                    csx={{
                      top: 0,
                      position: 'absolute',
                      margin: '0.625rem 0.25rem 0 0.5rem',
                    }}
                  />
                ) : (
                  <IconSearch
                    csx={{
                      color: 'base',
                      margin: '0.625rem 0.25rem 0 0.5rem',
                      top: 0,
                    }}
                  />
                )
              }
              csx={{
                height: '2.5rem',
                paddingY: '0.4375rem',
                paddingLeft: '2rem',
                paddingRight: '2.5rem',
                margin: 0,
              }}
              onChange={(e) => state.setValue(e.target.value)}
              buttonElements={
                state.showClear ? (
                  <Button
                    icon={<IconCancel />}
                    onClick={state.clear}
                    size="small"
                    variant="adaptative-dark"
                    csx={{
                      position: 'absolute',
                      top: 0,
                      right: 0,
                      marginRight: '0.375rem',
                    }}
                  />
                ) : undefined
              }
            />
            <VisuallyHidden>
              <button type="submit">Search</button>
            </VisuallyHidden>
          </Fragment>
        ),
      }
    },
  }
)

export interface SearchOptions {
  state: SearchFormState
}

export type SearchProps = ComponentPropsWithRef<typeof Search> & SearchOptions
