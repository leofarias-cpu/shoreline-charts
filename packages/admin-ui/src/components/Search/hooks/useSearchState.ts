import type { FormEvent } from 'react'
import { useState, useMemo, useCallback } from 'react'
import { useDebouncedCache } from '@vtex/admin-ui-hooks'

const DEFAULT_TIMEOUT_MS = 250

export function useSearchState(
  params: UseSearchStateParams = {
    initialValue: '',
    timeoutMs: DEFAULT_TIMEOUT_MS,
  }
): SearchFormState {
  const {
    initialValue = '',
    defaultValue = '',
    initiallyLoading = false,
    onSubmit: baseSubmit,
    timeoutMs = DEFAULT_TIMEOUT_MS,
  } = params

  const [value, debouncedValue, setValue] = useDebouncedCache({
    initialState: initialValue,
    timeoutMs,
  })

  const [loading, setLoading] = useState(initiallyLoading)

  const showClear = useMemo(() => value.toString().length > 0, [value])

  const onSubmit = useCallback((e: FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    baseSubmit?.()
  }, [])

  const clear = useCallback(() => {
    setValue(defaultValue)
  }, [value])

  return {
    value,
    debouncedValue,
    setValue,
    loading,
    setLoading,
    showClear,
    onSubmit,
    clear,
  }
}

export interface UseSearchStateParams {
  initialValue?: string
  defaultValue?: string
  initiallyLoading?: boolean
  onSubmit?: () => void
  timeoutMs?: number
}

export interface SearchFormState {
  value: string
  debouncedValue: string
  setValue: (value: string) => void
  onSubmit: (e: FormEvent<HTMLFormElement>) => void
  showClear: boolean
  loading: boolean
  setLoading: React.Dispatch<React.SetStateAction<boolean>>
  clear: () => void
}
