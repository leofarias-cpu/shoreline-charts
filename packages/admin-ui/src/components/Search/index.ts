export * from './Search'
export * from './state'
