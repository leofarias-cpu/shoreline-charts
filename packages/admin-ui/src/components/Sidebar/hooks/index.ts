export * from './useSidebarState'
export * from './useReducedState'
