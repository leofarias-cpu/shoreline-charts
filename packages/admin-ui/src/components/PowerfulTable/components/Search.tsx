import React from 'react'

import type { SearchProps } from '../../Search'
import { Search } from '../../Search'

export function TableSearch(props: SearchProps) {
  const { containerCsx, ...searchProps } = props

  return (
    <Search
      containerCsx={{ width: '25%', minWidth: '12.25rem', ...containerCsx }}
      {...searchProps}
    />
  )
}
