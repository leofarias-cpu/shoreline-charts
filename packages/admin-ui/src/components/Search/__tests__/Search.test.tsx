import React from 'react'
import { axe } from 'jest-axe'

import { Search } from '../Search'
import { useSearchState } from '../hooks/useSearchState'
import { render, withState } from '../../../test-utils'

const StatefulSearch = withState(Search, () => useSearchState())

const StatefulSearchLoading = withState(Search, () =>
  useSearchState({
    initiallyLoading: true,
  })
)

describe('Search tests', () => {
  it('should have overridable styles', () => {
    const { getByTestId } = render(
      <StatefulSearch
        data-testid="search"
        id="search"
        placeholder="Placeholder"
        aria-label="Product"
        csx={{ bg: 'azure' }}
      />
    )

    expect(getByTestId('search')).toHaveStyleRule('background', 'azure')
  })

  it('should match snapshot', () => {
    const { asFragment } = render(<StatefulSearch aria-label="Product" />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot on loading', () => {
    const { asFragment } = render(
      <StatefulSearchLoading aria-label="Product" />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should not have a11y violations', async () => {
    const { container } = render(
      <StatefulSearch id="search" aria-label="Product" />
    )

    const results = await axe(container)

    expect(results).toHaveNoViolations()
  })

  it('should not have a11y violations on loading', async () => {
    const { container } = render(
      <StatefulSearchLoading id="Search" aria-label="Product" />
    )

    const results = await axe(container)

    expect(results).toHaveNoViolations()
  })
})
