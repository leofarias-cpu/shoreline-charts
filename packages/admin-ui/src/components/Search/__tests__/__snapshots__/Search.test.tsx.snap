// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search tests should match snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: VTEXTrustVF,-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  line-height: 1.25;
  font-variation-settings: 'wght' 92;
  font-size: 0.875rem;
}

.emotion-2 {
  font-family: VTEXTrustVF,-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'wght' 92;
  width: 100%;
  height: 2.5rem;
  border-style: solid;
  border-width: 1px;
  padding-left: 2rem;
  padding-right: 2.5rem;
  border-color: #C4C5CA;
  border-radius: 4px;
  background-color: inherit;
  margin-top: 0.25rem;
  margin-bottom: 0.25rem;
  font-size: 0.875rem;
  color: #323845;
  outline: 0;
  -webkit-transition: all 70ms cubic-bezier(0.2, 0.2, 0.38, 0.9);
  transition: all 70ms cubic-bezier(0.2, 0.2, 0.38, 0.9);
  padding-top: 0.4375rem;
  padding-bottom: 0.4375rem;
  margin: 0rem;
}

.emotion-2:hover {
  border-color: #323845;
}

.emotion-2:focus {
  border-color: #2953B2;
  box-shadow: 0 0 0 1px #2953B2;
}

.emotion-2:disabled {
  background-color: #F4F6FB;
  color: #898F9E;
}

.emotion-2:focus+label {
  -webkit-transform: translate(1px, 4px) scale(0.875);
  -moz-transform: translate(1px, 4px) scale(0.875);
  -ms-transform: translate(1px, 4px) scale(0.875);
  transform: translate(1px, 4px) scale(0.875);
}

.emotion-2:placeholder-shown:not(:focus)+label {
  padding-top: 0.25rem;
}

.emotion-2:not(:placeholder-shown)+label {
  -webkit-transform: translate(1px, 4px) scale(0.875);
  -moz-transform: translate(1px, 4px) scale(0.875);
  -ms-transform: translate(1px, 4px) scale(0.875);
  transform: translate(1px, 4px) scale(0.875);
}

.emotion-3 {
  position: relative;
}

.emotion-3 label {
  left: 44px;
}

.emotion-1 {
  min-width: 20px;
  min-height: 20px;
  width: 20px;
  height: 20px;
  color: #323845;
  top: 0rem;
  left: 0rem;
  margin-left: 0.75rem;
  margin-right: 0.75rem;
  margin-top: 14px;
  position: absolute;
  margin: 0.625rem 0.25rem 0 0.5rem;
}

.emotion-4 {
  position: relative;
}

<form
    aria-label=" Form"
    class="emotion-4"
    role="search"
  >
    <span
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
    >
      <label
        class="emotion-0"
      >
         Input
      </label>
    </span>
    <div
      class="emotion-3"
    >
      <svg
        class="emotion-1"
        fill="none"
        viewBox="0 0 24 24"
      >
        <path
          clip-rule="evenodd"
          d="M15.6921 16.7489C12.9329 18.9787 8.87883 18.8112 6.31415 16.2466C3.57033 13.5027 3.57033 9.05413 6.31415 6.31031C9.05798 3.56649 13.5066 3.56649 16.2504 6.31031C18.8151 8.87498 18.9825 12.9291 16.7528 15.6883L19.5366 18.4721C19.8295 18.765 19.8295 19.2398 19.5366 19.5327C19.2437 19.8256 18.7688 19.8256 18.4759 19.5327L15.6921 16.7489ZM7.37481 7.37097C9.53285 5.21293 13.0317 5.21293 15.1898 7.37097C17.3277 9.5089 17.3476 12.9628 15.2495 15.1252C15.2275 15.1428 15.2063 15.1617 15.1859 15.1821C15.1656 15.2024 15.1466 15.2237 15.1291 15.2457C12.9666 17.3438 9.51274 17.3238 7.37481 15.1859C5.21678 13.0279 5.21678 9.52901 7.37481 7.37097Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <input
        class="emotion-2"
        value=""
      />
    </div>
    <span
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
    >
      <button
        type="submit"
      >
        Search
      </button>
    </span>
  </form>
</DocumentFragment>
`;

exports[`Search tests should match snapshot on loading 1`] = `
<DocumentFragment>
  @keyframes animation-0 {
  100% {
    -webkit-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-1 {
  0% {
    stroke-dasharray: 1,150;
    stroke-dashoffset: 0;
  }

  50% {
    stroke-dasharray: 90,150;
    stroke-dashoffset: -35;
  }

  100% {
    stroke-dasharray: 90,150;
    stroke-dashoffset: -124;
  }
}

.emotion-0 {
  font-family: VTEXTrustVF,-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  line-height: 1.25;
  font-variation-settings: 'wght' 92;
  font-size: 0.875rem;
}

.emotion-3 {
  font-family: VTEXTrustVF,-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'wght' 92;
  width: 100%;
  height: 2.5rem;
  border-style: solid;
  border-width: 1px;
  padding-left: 2rem;
  padding-right: 2.5rem;
  border-color: #C4C5CA;
  border-radius: 4px;
  background-color: inherit;
  margin-top: 0.25rem;
  margin-bottom: 0.25rem;
  font-size: 0.875rem;
  color: #323845;
  outline: 0;
  -webkit-transition: all 70ms cubic-bezier(0.2, 0.2, 0.38, 0.9);
  transition: all 70ms cubic-bezier(0.2, 0.2, 0.38, 0.9);
  padding-top: 0.4375rem;
  padding-bottom: 0.4375rem;
  margin: 0rem;
}

.emotion-3:hover {
  border-color: #323845;
}

.emotion-3:focus {
  border-color: #2953B2;
  box-shadow: 0 0 0 1px #2953B2;
}

.emotion-3:disabled {
  background-color: #F4F6FB;
  color: #898F9E;
}

.emotion-3:focus+label {
  -webkit-transform: translate(1px, 4px) scale(0.875);
  -moz-transform: translate(1px, 4px) scale(0.875);
  -ms-transform: translate(1px, 4px) scale(0.875);
  transform: translate(1px, 4px) scale(0.875);
}

.emotion-3:placeholder-shown:not(:focus)+label {
  padding-top: 0.25rem;
}

.emotion-3:not(:placeholder-shown)+label {
  -webkit-transform: translate(1px, 4px) scale(0.875);
  -moz-transform: translate(1px, 4px) scale(0.875);
  -ms-transform: translate(1px, 4px) scale(0.875);
  transform: translate(1px, 4px) scale(0.875);
}

.emotion-4 {
  position: relative;
}

.emotion-4 label {
  left: 44px;
}

.emotion-5 {
  position: relative;
}

.emotion-2 {
  min-width: 20px;
  min-height: 20px;
  width: 20px;
  height: 20px;
  color: #323845;
  top: 0rem;
  left: 0rem;
  margin-left: 0.75rem;
  margin-right: 0.75rem;
  margin-top: 14px;
  position: absolute;
  -webkit-animation: animation-0 1.5s linear infinite;
  animation: animation-0 1.5s linear infinite;
  margin: 0.625rem 0.25rem 0 0.5rem;
}

.emotion-1 {
  fill: none;
  stroke: #2953B2;
  stroke-width: 5;
  stroke-linecap: round;
  -webkit-animation: animation-1 1s ease-in-out infinite;
  animation: animation-1 1s ease-in-out infinite;
}

<form
    aria-label=" Form"
    class="emotion-5"
    role="search"
  >
    <span
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
    >
      <label
        class="emotion-0"
      >
         Input
      </label>
    </span>
    <div
      class="emotion-4"
    >
      <svg
        class="emotion-2"
        fill="none"
        focusable="false"
        viewBox="0 0 50 50"
      >
        <circle
          class="emotion-1"
          cx="25"
          cy="25"
          r="20"
        />
      </svg>
      <input
        class="emotion-3"
        value=""
      />
    </div>
    <span
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
    >
      <button
        type="submit"
      >
        Search
      </button>
    </span>
  </form>
</DocumentFragment>
`;
